### PR TITLE
Boolean parser

### PR DESCRIPTION
**Pull request by nikolai-laevskii**
Is part of [this epic](https://github.com/github/c2c-actions-akvelon/issues/35).

- If string complies to YAML 1.2 "Core Schema" specification, it will always be parsed according to it.
- If number can be parsed from string, that number will be converted to boolean (throws error if parameter strict is set to true)
- If number cannot be parsed from string, string is converted directly to boolean (throws error if parameter strict is set to true)
- Real booleans are just passed along

### Usage
```typescript
parseBoolean('true') // -> true
parseBoolean('false') // -> false

parseBoolean(1) // -> Error
parseBoolean(1, false) // -> true
parseBoolean("0", false) // -> false
parseBoolean((999).toString(16), false) // -> true

parseBoolean('False') // -> false
parseBoolean('FALSE') // -> false

parseBoolean('fAlSe') // -> Error
parseBoolean('fAlSe', false) // -> true
```

### Parameters
- `value: string | boolean | number` - value to have boolean parsed from
- `strict?: boolean` `default: true` - **true** sets the same behaviour as in `core.getBooleanInput` (parses a value only if it complies with the YAML 1.2 "Core Schema" specification and throws error otherwise). If **false**, it will parse any string, boolean or number without errors. String will be converted to number (if possible), number will be converted to boolean. If string cannot be converted to number, it will be converted directly to boolean

### CODE REVIEW

File: packages/helpers/src/boolean-parser.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      `Value does not meet YAML 1.2 "Core Schema" specification: ${value}\n` +
+        "Support boolean string list: `'true' | 'True' | 'TRUE' | 'false' | 'False' | 'FALSE'`"
+    )
+  }
+
+  if (typeof value === 'string') {
+    if (/^-?0+(\.?0+)?$/.test(value)) return false
+  }
+
+  return !!value
```
Comment (dsame) [inline packages/helpers/src/boolean-parser.ts]: I'd use `Boolean(value)` in order to obey elsint rule https://eslint.org/docs/latest/rules/no-implicit-coercion

File: packages/helpers/src/boolean-parser.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  if (falseValue.includes(String(value))) return false
+
+  if (strict) {
+    throw new TypeError(
+      `Value does not meet YAML 1.2 "Core Schema" specification: ${value}\n` +
+        "Support boolean string list: `'true' | 'True' | 'TRUE' | 'false' | 'False' | 'FALSE'`"
+    )
+  }
+
+  if (typeof value === 'string') {
```
Comment (dsame) [inline packages/helpers/src/boolean-parser.ts:31]: I'd added `value. toLowerCase() === 'false'` and `value. toLowerCase() === 'true'` here

File: packages/helpers/src/boolean-parser.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    if (/^-?0+(\.?0+)?$/.test(value)) return false
```
[the remark this answers is not in the document]
Comment (dsame) [reply packages/helpers/src/boolean-parser.ts]: A quick test
```
t1=new Date();for (let i=0;i<100000000;i++){Number("10.0")};console.log(((new Date()).getTime() - t1.getTime())/1000)
```
vs

```
 t1=new Date();for (let i=0;i<100000000;i++){/^-?0+(\.?0+)?$/.test("0.0")};console.log(((new Date()).getTime() - t1.getTime())/1000)

```

showed the `Number()` is a way faster than the regexp is.

**Review by dsame — APPROVED** [review]
I left some unsignificant notes that does not impact the approved status but can improve the PR in my view.

File: packages/helpers/src/boolean-parser.ts (hunk 2 above)
Comment (nikolai-laevskii) [reply packages/helpers/src/boolean-parser.ts:31]: This case is already covered on lines 21 and 22:
```typescript
  if (trueValue.includes(String(value))) return true
  if (falseValue.includes(String(value))) return false
```
so this addition would have no impact

File: packages/helpers/src/boolean-parser.ts (hunk 1 above)
Comment (nikolai-laevskii) [reply packages/helpers/src/boolean-parser.ts]: No objections here

File: packages/helpers/src/boolean-parser.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nikolai-laevskii) [reply packages/helpers/src/boolean-parser.ts]: Yeah, good point, seems like I overengineered this part a bit.

**Pull request closed without merging**